Adds deity mode (READY FOR REVIEW)

Kearel opened the pull request:
<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You find a README and example file in .\html\changelogs\ for further instructions.
-->
Deity mode is basically Mask of Nar-Sie with a few tiny twists to it.
In short, you have a group of cultists and a deity. The cultists have no special powers to begin with, only a special connection to the deity. The deity has special powers, but can't do them outside of their domain. So they have to give the cultists powers to increase said domain, so they can affect more area, and so that their power grows, so they can get more abilities to give to the cultists, etc etc.

All in all, its a cycle.

What needs to be done:
- [x] Replace/remove coder sprites/debugging code.
- [x] Finish/add nar-sie features based on feedback.
- [x] Add at least one more form.
- [ ] Ensure code is to standards.

Kearel committed:
Adds deity mode and the first form, Nar-Sie.

Kearel committed:
Fixes span issue

AceVX commented:
Sounds like a heckin' good concept. Eager to see where this goes.

Coureere commented:
We must construct additional pylons!

Iamgoofball commented:
hey folks

salvage code from Hand of God at /tg/ this was literally just that gamemode

Datraen commented on code/game/antagonist/station/highlander.dm:
```diff
@@ -0,0 +1,68 @@
+var/datum/antagonist/highlander/highlanders
```
Was this an error when rebasing to current code?

Kearel replied:
Yes.

atlantiscze commented on code/game/gamemodes/godmode/god_structures.dm:
```diff
@@ -0,0 +1,119 @@
+		src.visible_message("\The [src] crumbles!")
+		qdel(src)
+
+/obj/structure/deity/proc/attack_deity(var/mob/living/deity/deity)
```
It might be a good idea to define this on /obj/ level, for possible future expansion.

atlantiscze commented on code/game/gamemodes/godmode/god_turf.dm:
```diff
@@ -0,0 +1,23 @@
+/turf/simulated/floor/deity
```
Are these somehow removable barring RCD or "build wall remove wall" trick?

Kearel replied:
Depends on the flooring decl. But it should be, yes.

atlantiscze commented on code/modules/mob/living/deity/deity.dm:
```diff
@@ -0,0 +1,126 @@
+
+/mob/living/deity/verb/make_debug_structure()
+	set name = "Make structure"
+	set category = "CURRENTDEBUG"
```
I assume you plan to remove this at some point. If not, i'd suggest changing the category to standardized "Debug" so there's not another tab with single verb in it.

Kearel replied:
Yeah it was a debug tool used to test spawning structures and things like that.

atlantiscze commented on code/modules/mob/living/deity/deity_boons.dm:
```diff
@@ -0,0 +1,48 @@
+		A.forceMove(src)
+
+/mob/living/deity/proc/grant_boon(var/mob/living/L)
+	to_chat(L, "<span class='cult'>\The [src] grants you a boon of [current_boon]!</span>")
```
I would recommend moving this line down a bit, so it's behind these checks. Good place would be where the admin logs are generated.

atlantiscze commented on code/modules/mob/living/deity/deity_phenomena.dm:
```diff
@@ -0,0 +1,131 @@
+	if(!can_use_phenomena(20,L))
+		return
+
+	to_chat(L, "<span class='cult'><font size='2'>A weak power intrudes into your mind. You can choose to ignore it, or give in to servitude. <a href='?src=\ref[src];join_cult=1'>Give In</a></font></span>")
```
This would probably be better done with a yes/no prompt on client that, once clicked, notifies the deity of outcome. That way, following spawn(100) won't be needed, and in general the system will be more accurate (I can, for example, click the join button after 20 seconds).

In fact, the whole join_cult=1 feels weird to me. Is that actually being used now? I can pretty easily see how this could be hacked to join cult whenever you want, simply by calling the href manually.

Kearel replied:
Its actually how the current cult does it. Albeit with a rune.

atlantiscze replied:
My original point stands - this should probably be replaced with more secure (read: not href-hackable) method.

atlantiscze commented on code/modules/mob/living/deity/deity_phenomena.dm:
```diff
@@ -0,0 +1,131 @@
+		if(!is_follower(L))
+			to_chat(src, "<span class='notice'>\The [L] is resisting your power...</span>")
+
+/mob/living/deity/verb/conversion_forced(var/mob/living/L)
```
I can see some problems with forced conversion. I'd rather prefer to see a system that lets them resist though with possibility of eventual death or other serious drawbacks. This method should provide some benefits for the deity.

Kearel replied:
I dunno how exactly this would be done. Force conversion is already pretty hard though, as there is nothing actually keeping them there for that many seconds.

Maybe have it damage them if it fails?

atlantiscze reviewed (commented):
I am generally biased against cult in it's current form, and based on my subjective opinion this would be much more interesting alternative.

There are obvious balancing concerns with such large change, but that's very hard to fine tune without large scale testing, so i assume it will be tweaked in the future if necessary.

If you didn't do so already, i'd also suggest creating a wiki page with basic information.

Kearel committed:
More fixes

Kearel committed:
Fixes categories

Kearel committed:
Actually fixes category

Kearel committed:
Changes conversion methods. Regular is now a popup, forced can now be resisted.

Kearel commented:
Alright lemme know what you guys think about the new conversion method. Also any other suggestions?

Kearel committed:
Switches to defines for certain uplink unlocks. Particularly ones that unlock other things and categories.

Adds build shrine ability, which is basically how deity starts deitying and building structures.

Kearel committed:
A few modifications. Build Shrine can only be done on a turf with no dense objects. Removed debug verb.
Conversion only checks for altars. Adds helper proc for getting a type's name. Death stops conversion.

Kearel committed:
Fixes do_after and adds the projectile icons for blood drain. Looks okish.

Kearel committed:
Adds properly cropped icon of small-narsie for use for deity mode.

Iamgoofball commented:
@neerti

Kearel committed:
Yuuuuge changes. Switched schools on stuff, added start to wizard tower form. Remade phenomena. Started HUD stuff. Added blood stone icon.

MistakeNot4892 commented on code/datums/uplink/deity/form_specific/narsie/basic.dm:
```diff
@@ -0,0 +1,27 @@
+
+/datum/uplink_item/item/deity/boon/ability_dark/torment
+	name = "Torment"
+	desc = "Gives a follower the ability to cause mass hystery and pain."
```
`hysteria`

MistakeNot4892 commented on code/game/gamemodes/godmode/form_items/narsie_structures.dm:
```diff
@@ -0,0 +1,76 @@
+/obj/structure/deity/blood_forge
+	name = "unholy forge"
+	desc = "This forge gives off no heat, no light, its flames look almost unnaturual."
```
`unnatural`

MistakeNot4892 commented on code/game/gamemodes/godmode/form_items/narsie_items.dm:
```diff
@@ -0,0 +1,79 @@
+
+/obj/item/weapon/reagent_containers/food/drinks/zombiedrink
+	name = "well-used urn"
+	desc = "Said to bring those back to life, no matter the price."
```
`those` what?

MistakeNot4892 commented on code/game/gamemodes/godmode/form_items/narsie_structures.dm:
```diff
@@ -0,0 +1,76 @@
+	var/mob/living/carbon/human/H = user
+	user.visible_message("<span class='warning'>\The [user] calmly slices their finger on \the [src], smeering it over the black stone.</span>","<span class='warning'>You slowly slide your finger down one of \the [src]'s sharp edges, smeering it over its smooth surface.</span>")
+	do
+		user.audible_message("\The [user] utters something under their breathe.", "<span class='cult'>You mutter a dark prayer to your master as you feel the stone eat away at your lifeforce.</span>")
```
`breath`

MistakeNot4892 commented on code/game/gamemodes/godmode/god_altar.dm:
```diff
@@ -0,0 +1,96 @@
+
+	next_cycle = world.time + 10 SECONDS
+	cycles_before_converted--
+	if(!cycles_before_converted)
```
Isn't this going to tie straight back into the old cult furore about unwilling conversion?

Kearel replied:
They can resist it. They'll just die. Eventually.

MistakeNot4892 commented on code/modules/mob/living/deity/deity.dm:
```diff
@@ -0,0 +1,124 @@
+		eyeobj = new eye_type(src)
+		eyeobj.possess(src)
+		eyeobj.visualnet.add_source(src)
+	var/obj/item/uplink_holder = new(src) //because uplink devices need to be inside an item to work.
```
This seems like something of a hack. Would it be worth adjusting uplink code so it just has to be inside an atom, rather tan an obj/item?

Kearel replied:
It is definitely a hack. I'll try seeing if being inside a mob will actually break it.

MistakeNot4892 commented on code/modules/mob/living/deity/phenomena/conversion.dm:
```diff
@@ -0,0 +1,48 @@
+/datum/phenomena/forced_conversion/activate(var/mob/living/L)
+	var/obj/structure/deity/altar/A = locate() in get_turf(L)
+	A.set_target(L)
+	to_chat(src, "<span class='notice'>You embue \the [A] with your power, setting forth to force \the [L] to your will.</span>")
```
`imbue`

MistakeNot4892 commented on code/modules/mob/living/living_powers.dm:
```diff
@@ -15,3 +15,44 @@
+
+/mob/living/proc/breathe_death()
+	set name = "Breathe Death"
+	set desc = "Infect others with your very breathe."
```
`breath`

Kearel committed:
Finishes out wizard form. Finishes out icons and other things needed for completion.

Kearel committed:
merge dev

Kearel committed:
Adds blood projectile icon

Kearel committed:
Adds a category to an uplink item that did not have it.

PsiOmegaDelta commented on code/datums/uplink/deity/deity_items.dm:
```diff
@@ -0,0 +1,67 @@
+/datum/uplink_item/item/deity
```
The ```/datum/uplink_item/item``` is intended for physical items.
There is (or was) ```/datum/uplink_item/abstract``` for non-physical goods.

And why is this an antag uplink deal in the first place?

Kearel replied:
Alright yeah I can use that for the feat-based purchases.

It does the same thing for what I was going to do for deity purchases. I don't really see a reason for creating a whole new menu unless I have to.

PsiOmegaDelta commented on code/_onclick/hud/deity.dm:
```diff
@@ -0,0 +1,67 @@
+
+	ico = new(ui_style, "black")
+	ico.MapColors(0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0, -1,-1,-1,-1)
+	ico.DrawBox(rgb(255,255,255,1),1,ico.Height()/2,ico.Width()/2,ico.Height())
```
Icon operations are exceedingly undesirable.
Is there no way to do this using images instead?

Kearel replied:
To be honest I just C&Ped it from the other HUD stuff. I'll see if I can actually do away with it.

Kearel replied:
From what I can tell this is used so that you can click on the specific intents. I don't think I can actually remove this.

PsiOmegaDelta commented on code/game/antagonist/station/highlander.dm:
```diff
@@ -0,0 +1,66 @@
+		highlanders.add_antagonist(H.mind)
+
+	message_admins("<span class='notice'>[key_name_admin(usr)] used THERE CAN BE ONLY ONE!</span>", 1)
+	log_admin("[key_name(usr)] used there can be only one.")
```
Un-reported re-addition of the broken highlander mode. Has it been fixed?

PsiOmegaDelta commented on code/game/antagonist/station/highlander.dm:
```diff
@@ -0,0 +1,66 @@
+		return
+
+	for(var/mob/living/carbon/human/H in player_list)
+		if(H.stat == 2 || !(H.client)) continue
```
```if(H.stat == DEAD || !(H.client)) continue```

Kearel replied:
Honestly I have no idea why highlander shit keeps popping up. Probably something to do with a weird refactor of mine.

PsiOmegaDelta commented on code/game/objects/items/devices/uplink.dm:
```diff
@@ -31,13 +31,14 @@
 	var/next_offer_time					//The time a discount will next be offered
 	var/datum/uplink_item/discount_item	//The item to be discounted
 	var/discount_amount					//The amount as a percent the item will be discounted by
+	var/nano_state						//If set, use this instead of inventory state for nano_manager
```
Or just change 
```/obj/item/device/uplink/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)```
to 
```/obj/item/device/uplink/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1, state = inventory_state)```
and override the definition in new subtypes.

Kearel replied:
Yeah I could totally do that.

PsiOmegaDelta commented on code/modules/mob/living/deity/deity.dm:
```diff
@@ -0,0 +1,139 @@
+	structures.Cut()
+	qdel(eyeobj.visualnet)
+	qdel(eyeobj)
+	qdel(form)
```
```qdel_null()```

PsiOmegaDelta commented on code/modules/mob/living/deity/deity.dm:
```diff
@@ -0,0 +1,139 @@
+	qdel(form)
+	return ..()
+
+/mob/living/deity/verb/jump_to_follower()
```
No checks for death or otherwise?

PsiOmegaDelta commented on code/modules/mob/living/deity/deity_Stat.dm:
```diff
@@ -0,0 +1,17 @@
+/mob/living/deity/Stat()
+	. = ..()
+	statpanel("Status")
```
```
if(statpanel("Status"))
	[rest of stats here]
```

Perhaps even use a new tab.

PsiOmegaDelta commented on code/modules/mob/living/deity/deity_Stat.dm:
```diff
@@ -0,0 +1,17 @@
+/mob/living/deity/Stat()
+	. = ..()
+	statpanel("Status")
+	stat(null, "Health: [health]/[maxHealth]")
```
Should probably be ```stat("Health", "[health]/[maxHealth]")```.
Repeat for stats below.

PsiOmegaDelta commented on code/modules/mob/living/deity/deity_topic.dm:
```diff
@@ -0,0 +1,21 @@
+		var/atom/a = locate(href_list["pylon"])
+		if(a)
+			eyeobj.forceMove(get_turf(a))
+			to_chat(src, "<span class='notice'>Jumping to \the [a]</span>")
```
You're not returning 1 in handled hrefs.

Kearel replied:
Done.

PsiOmegaDelta commented on code/modules/mob/living/deity/deity_topic.dm:
```diff
@@ -0,0 +1,21 @@
+		return
+	if(href_list["intent"] && href_list["modifier"])
+		var/choice = input(src, "Set [href_list["intent"]]:[href_list["modifier"]] to:", "Phenomenas", null) as null|anything in phenomenas
+		if(!choice)
```
Need to check that the chosen phenomena is still in phenomenas by the time the user has deigned to answer.

Kearel replied:
Thats a good point.

PsiOmegaDelta commented on code/modules/mob/living/deity/forms.dm:
```diff
@@ -0,0 +1,102 @@
+	var/list/icon_states = list()
+	var/list/starting_feats //This is used to give different forms different starting trees
+
+/datum/god_form/New(var/mob/living/deity/D)
```
Call ```..()``` unless you know why you don't.

Kearel replied:
Done.

PsiOmegaDelta reviewed (changes requested):
Didn't make it all the way through, but here's a start.

PsiOmegaDelta commented on code/modules/mob/living/deity/deity.dm:
```diff
@@ -0,0 +1,139 @@
+	minions.Cut()
+	eyeobj.release()
+	structures.Cut()
+	qdel(eyeobj.visualnet)
```
It's the eyeobj's responsibility to qdel the visualnet.
One should rarely, if ever, qdel instances belonging to something else.

Kearel replied:
You're right. I'll make the eyeobj delete it instead.

Kearel committed:
Adds in psi suggestions outside of the HUD one. Unsure how to go about doing that.

Kearel committed:
Adds in the rest of Psi's suggestions. Remakes intent hud to use click position instead of making icons for it. Remakes uplink items to not use teh /item subtype.

Kearel commented:
Alright Psi I added in all of your suggestions.

comma commented on code/game/antagonist/outsider/deity.dm:
```diff
@@ -0,0 +1,20 @@
+var/datum/antagonist/deity/deity
```
why put it here instead of datum right below it? It won't be global or anything, just looks odd this way.

comma commented on code/game/antagonist/station/cult_god.dm:
```diff
@@ -0,0 +1,64 @@
+var/datum/antagonist/godcultist/godcult
```
Ditto

comma commented on code/game/gamemodes/godmode/god_altar.dm:
```diff
@@ -0,0 +1,96 @@
+	else ..()
+
+/obj/structure/deity/altar/process()
+	if(!target || world.time < next_cycle)
```
IMO could use a stat check for target here, to prevent conversion of KO'd people (and possible dead ones heh)

Kearel replied:
Hm... yeah that could work. It'll be harder to force a conversion, but I don't see anything wrong with that.

comma commented on code/game/gamemodes/godmode/god_structures.dm:
```diff
@@ -0,0 +1,95 @@
+	important_structure = 1
+	build_cost = 400
+	icon_state = "pylon"
+	var/datum/radio_frequency/radio_connection
```
why not just stick an intercomm object inside of it and use hear_talk to just pass a message to it?

Kearel replied:
I don't want it disabled due to an ion storm or something like that.

I don't know if there is a thing that I can use for that.

comma replied:
tl;dr of IRC stuff: then might as well just not depend on radio signals at all imo, go over deity's pylons in hear_talk to pass the message with audible_message. Would also make it work seamlessly with different deities.

comma commented on code/modules/mob/transform_procs.dm:
```diff
@@ -311,4 +311,19 @@
+	mutations |= CLUMSY //cause zombie
+	src.visible_message("<span class='danger'>\The [src]'s flesh decays before your very eyes!</span>", "<span class='danger'>Your entire body is ripe with pain as it is consumed down to flesh and bones. You... hunger.</span>")
+	if(src.mind)
+		src.mind.special_role = "Zombie"
```
Maybe put some sort of adminlog here, as they get antag status with this.

comma commented on code/modules/spells/aoe_turf/drain_blood.dm:
```diff
@@ -0,0 +1,66 @@
+			L.adjustToxLoss(-2.5)
+
+/obj/item/projectile/beam/blood_effect
+	name = "literally just blood"
```
I think this would look odd in ingame messages, 'Urist was hit by literally just blood!' 'Literally just blood misses you narrowly!'.
Maybe use something like 'blood jet' or spray or something along these lines.

comma commented on code/modules/spells/general/create_air.dm:
```diff
@@ -0,0 +1,25 @@
+/spell/create_air
+	name = "Create Air"
+	desc = "Use your mental strength to literally tear a hole from this dimension to the next, letting things through..."
```
Copypasta accident? Same description as Tear Veil, and kinda sinister for this spell :V I mean it might be accurate if you're stealing other dimension's oxygen but still.

comma commented:
re:zombies - need some hint messages as to what are they to do now. Are they to nom people? Are they to help culties? Are they to just be cool and harder to kill but otherwise carry on normally? Any hint would help.

Kearel committed:
Adds all of chinksy's suggestions. Bites the bullet and adds map change.

Kearel commented:
Done and done.

comma reviewed (approved):
good enough to toss at our free testers imo
